File: Andrei/Small00/Java/interpreter/util/SpecialFunction.java
package interpreter.util;

import java.util.Random;
import java.util.Scanner;

import interpreter.value.Value;
import interpreter.value.IntegerValue;
import interpreter.value.StringValue;

public class SpecialFunction extends Function {

    private FunctionType type;
    private Scanner in;
    private Random rand;

    public SpecialFunction(FunctionType type) {
        this.type = type;
        this.in = new Scanner(System.in);
        this.rand = new Random();
    }

    public Value<?> call(Instance self, Arguments args) {
        Value<?> v;

        switch (type) {
            case Print:
                v = this.print(args);
                break;
            case Println:
                v = this.println(args);
                break;
            case Read:
                v = this.read(args);
                break;
            case Random:
                v = this.random(args);
                break;
            // FIXME: add the others
            case Clone:
                v = this.clone(args);
                break;
            default:
                throw new RuntimeException("FIXME: implement me!");
        }

        return v;
    }

    private Value<?> print(Arguments args) {
        if (args.contains("arg1")) {
            Value<?> v = args.getValue("arg1");
            if (v instanceof IntegerValue) {
                IntegerValue iv = (IntegerValue) v;
                System.out.print(v.value());
            } else if (v instanceof StringValue) {
                StringValue sv = (StringValue) v;
                System.out.print(sv.value());
            } else {
                throw new RuntimeException("FIXME: Implement me!");
            }
        }

        return IntegerValue.Zero;
    }

    private Value<?> println(Arguments args) {
        Value<?> v = print(args);
        System.out.println();
        return v;
    }

    private Value<?> read(Arguments args) {
        // Print the argument.
        this.print(args);

        String str = in.nextLine();
        try {
           int n = Integer.parseInt(str);
           IntegerValue iv = new IntegerValue(n);
           return iv;
        } catch (Exception e) {
           StringValue sv = new StringValue(str);
           return sv;
        }
    }

    private Value<?> random(Arguments args) {
        if (!args.contains("arg1"))
            InterpreterError.abort("system.random: primeiro argumento inexistente");
        else if (!args.contains("arg2"))
            InterpreterError.abort("system.random: segundo argumento inexistente");

        Value<?> v1 = args.getValue("arg1");
        Value<?> v2 = args.getValue("arg2");

        if (!(v1 instanceof IntegerValue))
            InterpreterError.abort("system.random: primeiro argumento não é inteiro");
	else if (!(v2 instanceof IntegerValue))
            InterpreterError.abort("system.random: segundo argumento não é inteiro");

        int n1 = ((IntegerValue) v1).value();
        int n2 = ((IntegerValue) v2).value();

        if (n2 <= n1)
            InterpreterError.abort("system.random: segundo argumento menor/igual ao primeiro");

        int r = rand.nextInt(n2 - n1 + 1) + n1;
        IntegerValue iv = new IntegerValue(r);
        return iv;
    }

    private Value<?> clone(Arguments args) {
        if (!args.contains("arg1"))
            InterpreterError.abort("clone: primeiro argumento inexistente");

        Value<?> v = args.getValue("arg1");
        if (!(v instanceof InstanceValue))
            InterpreterError.abort("clone: primeiro argumento não é instance");

        Instance i = ((InstanceValue) v).value();
        InstanceValue iv = new InstanceValue(i.dup());
        return iv;
    }

}

File: Andrei/Small00/C++/interpreter/util/SpecialFunction.cpp
#include "SpecialFunction.h"

#include <ctime>
#include <cstdio>
#include <string>
#include <cstdlib>
#include <iostream>

#include "Arguments.h"
#include "InterpreterError.h"
#include "../value/StringValue.h"
#include "../value/IntegerValue.h"
#include "../value/InstanceValue.h"
#include "../value/FunctionValue.h"

SpecialFunction::SpecialFunction(FunctionType type) : m_type(type) {
    srand(time(0));
}

SpecialFunction::~SpecialFunction() {
}

Type* SpecialFunction::call(Instance* self, Arguments* args) {
    Type* value;

    switch (m_type) {
        case PrintFunction:
            value = this->print(args);
            break;
        case PrintlnFunction:
            value = this->println(args);
            break;
        case ReadFunction:
            value = this->read(args);
            break;
        case RandomFunction:
            value = this->random(args);
            break;
        case CloneFunction:
            value = this->clone(args);
            break;
        default:
            throw std::string("FIXME: Implement me!");
    }

    return value;
}

Type* SpecialFunction::print(Arguments* args) {
    if (args->contains("arg1")) {
        Type* v = args->value("arg1");
        if (v->type() == Type::IntegerType) {
            IntegerValue* iv = (IntegerValue*) v;
            printf("%d", iv->value());
        } else if (v->type() == Type::StringType) {
            StringValue* sv = (StringValue*) v;
            printf("%s", sv->value().c_str());
        } else {
            throw std::string("FIXME: Implement me!");
        }
    }

    return IntegerValue::Zero;
}

Type* SpecialFunction::println(Arguments* args) {
    Type* v = this->print(args);
    printf("\n");
    return v;
}

Type* SpecialFunction::read(Arguments* args) {
    // Print the argument.
    Type* v = this->print(args);

    std::string input;
    std::cin >> input;

    char* valid;
    int n = (int) strtol(input.c_str(), &valid, 10);

    if (*valid == 0) {
        IntegerValue* iv = new IntegerValue(n);
        return iv;
    } else {
        StringValue* sv = new StringValue(input);
        return sv;
    }

    return v;
}

Type* SpecialFunction::random(Arguments* args) {
    if (!args->contains("arg1"))
        InterpreterError::abort("random: primeiro argumento inexistente");
    else if (!args->contains("arg2"))
        InterpreterError::abort("random: segundo argumento inexistente");

    Type* v1 = args->value("arg1");
    Type* v2 = args->value("arg2");

    if (v1->type() != Type::IntegerType)
        InterpreterError::abort("random: primeiro argumento nao eh inteiro");
    else if (v2->type() != Type::IntegerType)
        InterpreterError::abort("random: segundo argumento nao eh inteiro");

    int n1 = ((IntegerValue*) v1)->value();
    int n2 = ((IntegerValue*) v2)->value();

    if (n2 <= n1)
        InterpreterError::abort("random: segundo argumento menor/igual ao primeiro");

    int r = (rand() % (n2 - n1 + 1)) + n1;
    IntegerValue* iv = new IntegerValue(r);
    return iv;
}

Type* SpecialFunction::clone(Arguments* args) {
    if (!args->contains("arg1"))
        InterpreterError::abort("clone: primeiro argumento inexistente");

    Type* v = args->value("arg1");
    if (v->type() != Type::InstanceType)
        InterpreterError::abort("clone: primeiro argumento nao eh instancia");

    Instance* i = ((InstanceValue*) v)->value();
    InstanceValue* iv = new InstanceValue(i->dup());
    return iv;
}


